Feature/sn 850 ephemeral storage solution

namnguyen-saritasa opened the pull request:
Solutions explored:
- `Microsoft.EntityFrameworkCore.InMemory`: As explained in the ticket, this would leak memory.
- In memory SQLite: Both this and the previous one have the same problem: Every time a query/save changes is executed, `RelationalConnection.CreateDbConnection` is invoked, creating a new connection. And since it is nigh impossible to know which connection is associated with which session without manual book-keeping (we would have to expose the hidden `sqlite3_serialize` native function, serialize the DB's content, then store it somewhere and manually deserialize it when `RelationalConnection.CreateDbConnection` is invoked), this solution is not viable.

I have created a custom ephemeral SQLite implementation, which would create a disk-based SQLite store when a scope is created, and delete it when a scope ends. Since the file is tracked by the scope itself, no additional book-keeping is required.

I have also implemented migration and seeding for this new connector type. The basic idea is that we create a new SQLite instance when the server is started, migrate and seed it, then copy its content over when a new scoped DB instance is created.

Note: per discussion w/ @vasiliy-chefonov, I will not be implementing S3 object expiration in this PR.

namnguyen-saritasa committed:
Added ephemeral sqlite SN-850

namnguyen-saritasa committed:
WIP seeding

namnguyen-saritasa committed:
Added ephemeral sqlite SN-850

namnguyen-saritasa committed:
Automatically Seed data SN-850

namnguyen-saritasa committed:
Refactored SN-850

namnguyen-saritasa committed:
Removed junks SN-850

qqeekk commented:
> I have created a custom ephemeral SQLite implementation, which would create a disk-based SQLite store when a scope is created, and delete it when a scope ends.

So each request will be addressed to a separate instance?
Can we create a singleton session storage to track and create a session each time user logins to the system and dispose it when they log out? Within a session - all requests will be addressed to the same database instance

namnguyen-saritasa commented:
> Can we create a singleton session storage to track and create a session each time user logins to the system and dispose it when they log out? Within a session - all requests will be addressed to the same database instance

I would not recommend this approach, since it would break if we have more than one demo server. We can use a central database to keep track of these session (redis should work well here), but it would severely overcomplicate an otherwise simple demo. We would also have to handle cases in which a user logged in but would not log out.

Plus, this is beyond the scope of this task

AlexanderMakarov124 commented on demo/Saritasa.NetForge.Demo/Infrastructure/Storage/EphemeralSqliteConnectionFactory.cs:
```diff
@@ -0,0 +1,78 @@
+/// </remarks>
+internal class EphemeralSqliteConnectionFactory : CriticalFinalizerObject, IEphemeralSqliteConnectionFactory
+{
+    private readonly string path = Path.GetTempFileName();
```
Consider applying fix like in [here](https://cloud-ci.sgs.com/sonar/coding_rules?open=csharpsquid%3AS5445&rule_key=csharpsquid%3AS5445)

I recommend `SonarQube` extension to see such warnings in the IDE

Also, what does mean `path`? Path to what?

namnguyen-saritasa replied:
I have made temp file generation more secured

> Also, what does mean `path`? Path to what?

Please refer to the PR's description for such questions

AlexanderMakarov124 commented on demo/Saritasa.NetForge.Demo/Infrastructure/Storage/EphemeralSqliteConnectionFactory.cs:
```diff
@@ -0,0 +1,78 @@
+
+    private static async Task BackupDatabase(string targetPath, string destinationPath, CancellationToken cancellationToken)
+    {
+        await using var targetConn = CreateConnection(targetPath);
```
Use full name `targetConnection` instead of `targetConn`. It's bad practice to use contractions

AlexanderMakarov124 commented on demo/Saritasa.NetForge.Demo/Infrastructure/Storage/EphemeralSqliteExtensions.cs:
```diff
@@ -0,0 +1,22 @@
+        var extension = new EphemeralSqliteOptionsExtension();
+        ((IDbContextOptionsBuilderInfrastructure)optionsBuilder).AddOrUpdateExtension(extension);
+        // Connection string does not matter since we will be overriding the connection provider
+        return optionsBuilder.UseSqlite(string.Empty);
```
Prefer using parameter name when it isn't obvious for what parameter some value was used. Also, we usually use parameter names with `true`, `false` and `null`
```suggestion
        return optionsBuilder.UseSqlite(connectionString: string.Empty);
```

AlexanderMakarov124 commented on demo/Saritasa.NetForge.Demo/Infrastructure/Storage/EphemeralSqliteConnectionFactory.cs:
```diff
@@ -0,0 +1,78 @@
+
+    /// <inheritdoc cref="Dispose()" />
+    /// <param name="disposing">Whether this method is called by <see cref="Dispose()"/> or not.</param>
+    protected virtual void Dispose(bool disposing)
```
Is it fine that `disposing` isn't used?

namnguyen-saritasa replied:
This is a virtual method, we might not use this value in this class, but we might use it for derived class. This is standard procedure.

AlexanderMakarov124 commented on demo/Saritasa.NetForge.Demo/Infrastructure/Storage/DbSnapshot.cs:
```diff
@@ -0,0 +1,44 @@
+/// Stores the database snapshot's location.
+/// Automatically delete it after the application is closed.
+/// </summary>
+internal class DbSnapshot : IDisposable
```
Can we implement disposed pattern here too? Like in `EphemeralSqliteConnectionFactory`

AlexanderMakarov124 commented on demo/Saritasa.NetForge.Demo/Infrastructure/Middlewares/DbSnapshotLoaderMiddleware.cs:
```diff
@@ -0,0 +1,39 @@
+    {
+        var sp = httpContext.RequestServices;
+        var snapshot = sp.GetRequiredService<DbSnapshot>();
+        ArgumentNullException.ThrowIfNull(snapshot.SnapshotLocation, nameof(snapshot.SnapshotLocation));
```
Looks like when you remove second parameter, the name of first one will be used, so you can remove it

AlexanderMakarov124 commented on demo/Saritasa.NetForge.Demo/Infrastructure/Middlewares/DbSnapshotLoaderMiddleware.cs:
```diff
@@ -0,0 +1,39 @@
+    /// <param name="httpContext">HTTP context.</param>
+    public async Task Invoke(HttpContext httpContext)
+    {
+        var sp = httpContext.RequestServices;
```
Please use full names for variables in this method

AlexanderMakarov124 commented on demo/Saritasa.NetForge.Demo/Infrastructure/Middlewares/DbSnapshotLoaderMiddleware.cs:
```diff
@@ -0,0 +1,39 @@
+        var dbContext = sp.GetRequiredService<ShopDbContext>();
+        IInfrastructure<IServiceProvider> infrastructure = dbContext.Database;
+
+        var ephemeralConnection = infrastructure.Instance.GetRequiredService<IEphemeralSqliteConnectionFactory>();
```
We have the same code in `DatabaseInitializer.DumpDatabase` to get ephemeral connection, what do you think about moving it to a separate method? It will allow you to leave one comment to it, because when I first saw this code in this place, I didn't understand why we use `Infrastructure` instead of regular service provider

AlexanderMakarov124 commented on demo/Saritasa.NetForge.Demo/Infrastructure/Seeders/UsersSeeder.cs:
```diff
@@ -7,8 +7,9 @@ namespace Saritasa.NetForge.Demo.Infrastructure.Seeders;
-internal class UsersSeeder
+internal class UsersSeeder : ISeeder
 {
+    private const string DefaultPassword = "11111111Aa";
```
We usually use `Qwerty123#` as default password

namnguyen-saritasa replied:
It was not before this PR: 5150424bd3e41445de110e66ab312afea953291b

AlexanderMakarov124 commented on demo/Saritasa.NetForge.Demo/Infrastructure/Startup/DatabaseInitializer.cs:
```diff
@@ -8,20 +11,44 @@ namespace Saritasa.NetForge.Demo.Infrastructure.Startup;
     /// <param name="dbContext">Data context.</param>
-    public DatabaseInitializer(ShopDbContext dbContext)
+    /// <param name="dbSnapshot">Database snapshot.</param>
+    /// <param name="seeders">Seedeers.</param>
```
```suggestion
    /// <param name="seeders">Seeders.</param>
```

AlexanderMakarov124 commented on demo/Saritasa.NetForge.Demo/Program.cs:
```diff
@@ -1,5 +1,7 @@
 using McMaster.Extensions.CommandLineUtils;
+using Microsoft.Data.Sqlite;
```
Redundant usings?

AlexanderMakarov124 commented on demo/Saritasa.NetForge.Demo/Saritasa.NetForge.Demo.csproj:
```diff
@@ -19,6 +19,7 @@
           <PrivateAssets>all</PrivateAssets>
           <IncludeAssets>runtime; build; native; contentfiles; analyzers; buildtransitive</IncludeAssets>
         </PackageReference>
+        <PackageReference Include="Microsoft.EntityFrameworkCore.Sqlite" Version="8.0.0" />
```
Do we need to remove `PostgreSQL` related references?

namnguyen-saritasa replied:
We still use them in Designers file. We could remove all migrations and re-migrate though

AlexanderMakarov124 reviewed (changes requested):
I cannot access file because it's used by another process. I see this error when tried to launch app in debug locally. Did you have problems with it?

namnguyen-saritasa committed:
Applied fixes SN-850

namnguyen-saritasa committed:
Removed redundant parameter SN-850

namnguyen-saritasa commented:
> I cannot access file because it's used by another process. I see this error when tried to launch app in debug locally. Did you have problems with it?

I did not encounter such error for both Debug and Release build.
Tested on: macOS (arm64), Linux (x86_64)

EDIT: I was able to replicate this on Windows so I will draft this PR to investigate further

namnguyen-saritasa committed:
Disabled pooling SN-850

namnguyen-saritasa commented:
> I cannot access file because it's used by another process. I see this error when tried to launch app in debug locally. Did you have problems with it?

By default, connections are [pooled](https://learn.microsoft.com/en-us/ef/core/what-is-new/ef-core-6.0/breaking-changes#connection-pool), so even after an `SqliteConnection` object is finalized, it would not close the file handle. Since every session create a new database, this behavior is not needed and have been disabled.

AlexanderMakarov124 commented on demo/Saritasa.NetForge.Demo/Infrastructure/Startup/DatabaseInitializer.cs:
```diff
@@ -33,9 +34,10 @@ private async Task<string> DumpDatabase(CancellationToken cancellationToken)
     {
         // dbContext uses a service provider different from the one
```
Move this comment to `InfrastructureExtensions.GetInstance`

namnguyen-saritasa committed:
Moved remark to a more appropriate location SN-850

vasiliy-chefonov commented on demo/Saritasa.NetForge.Demo/Startup.cs:
```diff
@@ -1,10 +1,14 @@
 ﻿using System.Globalization;
 using Microsoft.AspNetCore.Identity;
 using Microsoft.EntityFrameworkCore;
+using Microsoft.Extensions.DependencyInjection.Extensions;
```
Redundant using

vasiliy-chefonov commented on demo/Saritasa.NetForge.Demo/Infrastructure/DependencyInjection/EphemeralSqliteModule.cs:
```diff
@@ -0,0 +1,36 @@
+    public static void AddEphemeralSqlite(this IServiceCollection services)
+    {
+        services.AddHttpContextAccessor();
+        services.TryAddSingleton<DbSnapshot>();
```
Could you please comment on why we need to reuse this snapshot?

vasiliy-chefonov commented on demo/Saritasa.NetForge.Demo/Infrastructure/Storage/DbSnapshot.cs:
```diff
@@ -0,0 +1,59 @@
+        set
+        {
+            snapShotLocation = value;
+            logger.LogInformation("Snapshot location set: {Location}", value);
```
Can it be on the Debug level?

vasiliy-chefonov commented on demo/Saritasa.NetForge.Demo/Startup.cs:
```diff
@@ -35,16 +39,14 @@ public Startup(IConfiguration configuration)
     public void ConfigureServices(IServiceCollection services, IWebHostEnvironment environment)
     {
         services.AddRazorPages();
-        var connectionString = configuration.GetConnectionString("AppDatabase")
```
Do we have to update the settings template?

vasiliy-chefonov commented:
I got this while trying to add the new shop.

<img width="866" alt="image" src="https://github.com/user-attachments/assets/31d03c89-2b8c-4c03-9a7b-d0e139e9fc65" />

AlexanderMakarov124 commented on demo/Saritasa.NetForge.Demo/Infrastructure/Storage/EphemeralSqliteConnectionFactory.cs:
```diff
@@ -0,0 +1,78 @@
+        GC.SuppressFinalize(this);
+    }
+
+    private static async Task BackupDatabase(string targetPath, string destinationPath, CancellationToken cancellationToken)
```
Can we rename `targetPath` to `sourcePath`? Target and destination sounds a little similar

namnguyen-saritasa committed:
Fixed db early disposal problem SN-850

namnguyen-saritasa committed:
Fixed service registration SN-850

namnguyen-saritasa committed:
Updated settings template SN-850

namnguyen-saritasa committed:
Added manual connection Cancellation SN-850

namnguyen-saritasa committed:
Refactored parameter name SN-850

namnguyen-saritasa committed:
Refactored parameter name SN-850

namnguyen-saritasa committed:
Merge branch 'develop' into feature/SN-850-ephemeral-storage-solution

# Conflicts:
#	demo/Saritasa.NetForge.Demo/Saritasa.NetForge.Demo.csproj

AlexanderMakarov124 reviewed (changes requested):
I have some error related to migrations right now when launching the application. Do you have the same?

![image](https://github.com/user-attachments/assets/0d1e6a2c-6f32-4759-bb7c-d5394f4df62a)

namnguyen-saritasa committed:
Removed old PostgreSQL dependencies SN-850

namnguyen-saritasa committed:
Re-added `DatabaseGenerated` SN-850

namnguyen-saritasa commented:
@AlexanderMakarov124 I have fixed this problem by removing the old PostgreSQL migrations. Also, SQLite does not support `DatabaseGenerated` so I have added an initial value to it.

namnguyen-saritasa committed:
Removed command line utils SN-850

vasiliy-chefonov commented on demo/Saritasa.NetForge.Demo/Infrastructure/Startup/DatabaseInitializer.cs:
```diff
@@ -1,5 +1,9 @@
-﻿using Extensions.Hosting.AsyncInitialization;
+using Extensions.Hosting.AsyncInitialization;
 using Microsoft.EntityFrameworkCore;
+using Microsoft.EntityFrameworkCore.Infrastructure;
```
Redundant using

vasiliy-chefonov commented on demo/Saritasa.NetForge.Demo/Infrastructure/Storage/EphemeralSqliteConnectionFactory.cs:
```diff
@@ -0,0 +1,108 @@
+    
+    private static string GetNewDbPath() => FileHelpers.CreateTemporaryFileSecure(".sqlite");
+
+    public EphemeralSqliteConnectionFactory(IHttpContextAccessor httpContextAccessor)
```
Missing docs

vasiliy-chefonov commented on demo/Saritasa.NetForge.Demo/Infrastructure/Middlewares/DbSnapshotLoaderMiddleware.cs:
```diff
@@ -0,0 +1,56 @@
+using Microsoft.EntityFrameworkCore.Infrastructure;
```
Redundant using

vasiliy-chefonov commented on demo/Saritasa.NetForge.Demo/Infrastructure/Middlewares/DbSnapshotLoaderMiddleware.cs:
```diff
@@ -0,0 +1,56 @@
+        using var linkedCancellationTokenSource = CancellationTokenSource.CreateLinkedTokenSource(cancellationToken,
+            manualCancellationTokenSource.Token);
+        cancellationToken = linkedCancellationTokenSource.Token;
+        httpContext.RequestAborted = cancellationToken;
```
I guess overwriting `HttpContext.RequestAborted` might break cancellation in earlier middleware that registered to the original token. Even with a linked token, those registrations won’t carry over. Safer to pass the linked token explicitly instead of modifying the context.

vasiliy-chefonov commented on demo/Saritasa.NetForge.Demo/Infrastructure/Storage/EphemeralSqliteConnectionFactory.cs:
```diff
@@ -0,0 +1,108 @@
+            return;
+        }
+
+        var cancellationToken = httpContext.RequestAborted;
```
I guess we can unite the first clause with this one. Could you please also comment on why we check for `httpContext` existence and cancellation possibility?

vasiliy-chefonov commented on demo/Saritasa.NetForge.Demo/Infrastructure/Storage/EphemeralSqliteConnectionFactory.cs:
```diff
@@ -0,0 +1,108 @@
+
+    public EphemeralSqliteConnectionFactory(IHttpContextAccessor httpContextAccessor)
+    {
+        var httpContext = httpContextAccessor.HttpContext;
```
Do we really need to rely on http context here? Can we pass a flag instead (whether it's a scoped request or not)?

namnguyen-saritasa replied:
We need to rely on `HttpContext` to bind a single database to a session, since a session could have an arbitrary amount of scopes, this makes sure all components are using the same database

namnguyen-saritasa replied:
Although technically still depends on HTTP context, I have made the session associated with a manual cancellation source instead of an `HttpContext` object

vasiliy-chefonov commented on demo/Saritasa.NetForge.Demo/Saritasa.NetForge.Demo.csproj:
Do we still need `AspNetCore.HealthChecks.NpgSql`?

namnguyen-saritasa replied:
No, I will remove it

namnguyen-saritasa committed:
Applied fixes SN-850

vasiliy-chefonov commented on demo/Saritasa.NetForge.Demo/Infrastructure/Storage/EphemeralSqliteConnectionFactory.cs:
```diff
@@ -12,7 +12,7 @@ namespace Saritasa.NetForge.Demo.Infrastructure.Storage;
 internal class EphemeralSqliteConnectionFactory : CriticalFinalizerObject, IEphemeralSqliteConnectionFactory
 {
-    private static ConcurrentDictionary<HttpContext, string> SqliteDbPathStores { get; } = new();
+    private static ConcurrentDictionary<CancellationToken, string> SqliteDbPathStores { get; } = new();
```
Tokens are meant for cancellation control, not identity.

It’s not intuitive why injecting `IManualCancellationTokenAccessor` should determine which file we’re using. It violates the separation of concerns.

vasiliy-chefonov commented on demo/Saritasa.NetForge.Demo/Infrastructure/Storage/EphemeralSqliteConnectionFactory.cs:
```diff
@@ -0,0 +1,111 @@
+    /// <summary>
+    /// Constructor.
+    /// </summary>
+    public EphemeralSqliteConnectionFactory(IManualCancellationTokenAccessor manualCancellationTokenAccessor)
```
Is it possible to store the unique session ID associated with the HttpContext?

Like we can set in the middleware before the request:
`httpContext.Items["EphemeralSqliteSessionId"] = Guid.NewGuid();`

And later we can use it here:
`var sessionId = httpContextAccessor.HttpContext?.Items["EphemeralSqliteSessionId"] as Guid?;`

So, later we can operate with IDs instead.

namnguyen-saritasa committed:
Use GUID to identify HTTP sessions SN-850
